{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "B2B Tools - MSA Example",
      "private_outputs": true,
      "provenance": [],
      "collapsed_sections": [],
      "authorship_tag": "ABX9TyO/w5+jEoURIdaNVqyfSZ04",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/agdiaz/Bloom/blob/master/Bio2ByteTools_v3_multipleseq_pypi.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "wzSJQsPWYV8Z"
      },
      "outputs": [],
      "source": [
        "#@title 1. Install the Bio2Byte tools package and its dependencies\n",
        "\n",
        "#@markdown Once this cell has been executed, a zip-archive with\n",
        "#@markdown the obtained prediction will be automatically downloaded\n",
        "#@markdown to your computer.\n",
        "\n",
        "#@markdown **Important**: Given we are installing new versions of some dependencies\n",
        "#@markdown please click on the \"Restart runtime\" button after pip finished \n",
        "#@markdown installing libraries (and run this cell again)\n",
        "\n",
        "!pip install b2bTools==3.0.4 biopython matplotlib==3.4.0 ipympl"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 2.a Upload your MSA file\n",
        "\n",
        "#@markdown Once this cell has been executed, your target MSA file\n",
        "#@markdown will be ready to run the predictions on it\n",
        "import os\n",
        "import shutil\n",
        "from google.colab import files\n",
        "\n",
        "if os.path.exists(\"/content/results\"):\n",
        "    shutil.rmtree(\"/content/results\")\n",
        "\n",
        "msa_filename = None\n",
        "\n",
        "uploaded = files.upload()\n",
        "\n",
        "for fn in uploaded.keys():\n",
        "  print('MSA file \"{name}\" with length {length} bytes uploaded with success'.format(\n",
        "      name=fn, length=len(uploaded[fn])))\n",
        "  msa_filename = fn"
      ],
      "metadata": {
        "cellView": "form",
        "id": "7p6nFboEZHr2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 2.b MSA file format\n",
        "#@markdown This cell lists the file formats this notebook can read\n",
        "\n",
        "msa_type = 'fasta' #@param [\"clustal\", \"fasta\", \"phylip\", \"stockholm\", \"emboss\"]"
      ],
      "metadata": {
        "cellView": "form",
        "id": "uasy39UNbZFI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 3. Get the predictions for each sequence of the alignment\n",
        "\n",
        "#@markdown Once this cell has been executed, the Bio2Byte tools predict\n",
        "#@markdown the values sequence by sequences of the alignment\n",
        "%%capture\n",
        "\n",
        "from b2bTools.multipleSeq.Predictor import MineSuiteMSA\n",
        "\n",
        "msaSuite = MineSuiteMSA()\n",
        "msaSuite.predictAndMapSeqsFromMSA(f\"/content/{msa_filename}\", predTypes = ('eFoldMine', 'disoMine', 'dynamine'))\n",
        "msaSuite.getDistributions()"
      ],
      "metadata": {
        "cellView": "form",
        "id": "GRjKv52KaLRu"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 4. Sequence conservation (Shannon's entropy)\n",
        "#@markdown These functions declared here will be used in the following cells to\n",
        "#@markdown plot different predictions\n",
        "\n",
        "import math\n",
        "import Bio\n",
        "from Bio import AlignIO\n",
        "\n",
        "#Shannon_entropy calculates the conservation of the amino acids at each position. It takes into account \n",
        "#the gaps in the columns. Check https://doi.org/10.1002/prot.10146 for info formula\n",
        "def shannon_entropy(list_input):\n",
        "    tot = len(list_input) #total number of AA at particular position in MSA\n",
        "    gaps = list_input.count(\"-\") / tot #count number of gaps in that column of MSA\n",
        "    unique_base = set(list_input) #remove duplicates\n",
        "    unique_base_len = len(unique_base) #total number of AA at particular position in MSA with no duplicates\n",
        "    \n",
        "    entropy_list = [] # entropy of AA at particular position\n",
        "    \n",
        "    for base in unique_base:\n",
        "        n_i = list_input.count(base)                 \n",
        "        P_i = n_i / tot\n",
        "        entropy_i = P_i * (math.log(P_i, 2))\n",
        "        \n",
        "        entropy_list.append(entropy_i)\n",
        "    \n",
        "    #sum entropy of every residue at a position and normalize it so it is between 0 and 1\n",
        "    entropy_sum = math.fsum(entropy_list)\n",
        "    min_length = min(tot, unique_base_len)\n",
        "    \n",
        "    if min_length == 1: # log(1, 2) = 0; n/0 throws ZeroDivisionError\n",
        "        shannon_entropy = entropy_sum\n",
        "    else:\n",
        "        length_log = math.log(min_length, 2)\n",
        "        shannon_entropy = (-1 / length_log) * entropy_sum\n",
        "    \n",
        "    #Return entropy AA and entropy gaps at 1 position\n",
        "    #If entropy is high than there are many possible arrangements (high variability)\n",
        "    return shannon_entropy, gaps\n",
        "\n",
        "\n",
        "def conservation(alignment_file):\n",
        "    conservation_AA_list = []\n",
        "\n",
        "    for col_no in range(len(list(alignment_file[0]))):\n",
        "        list_input = list(alignment_file[:, col_no])\n",
        "\n",
        "        sh_entropy_AA, sh_entropy_gaps = shannon_entropy(list_input)\n",
        " \n",
        "        #Translate entropy into conservation. 0 means no conservation, 1 highly conserved\n",
        "        conservation_AA = (1 - sh_entropy_AA) * (1 - sh_entropy_gaps)\n",
        "        \n",
        "        conservation_AA_list.append(conservation_AA)\n",
        "\n",
        "    return conservation_AA_list\n",
        "\n",
        "#Read MSA\n",
        "alignment_file = AlignIO.read(msa_filename, msa_type)\n",
        "conservation_AA_list = conservation(alignment_file)"
      ],
      "metadata": {
        "cellView": "form",
        "id": "DDZ7JQfqaMhi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 5. Prepare to plot results\n",
        "#@markdown Run this cell in order to prepare the notebook context to render different plots\n",
        "\n",
        "from google.colab import output\n",
        "output.enable_custom_widget_manager()\n",
        "%matplotlib widget\n",
        "\n",
        "import matplotlib.pyplot as plt\n",
        "import os\n",
        "\n",
        "if not os.path.exists(\"/content/results\"):\n",
        "    os.mkdir(\"/content/results\", )\n",
        "\n",
        "PREDICTION_TITLES = {\n",
        "    'backbone': \"DynaMine backbone dynamics\",\n",
        "    'sidechain': \"DynaMine sidechain dynamics\",\n",
        "    'sheet': \"DynaMine conformational propensities: Sheet\",\n",
        "    'coil': \"DynaMine conformational propensities: Coil\",\n",
        "    'helix': \"DynaMine conformational propensities: Helix\",\n",
        "    'ppII': \"DynaMine conformational propensities: ppII (polyproline II)\",\n",
        "    'disoMine': \"Disorder (disoMine)\",\n",
        "    'earlyFolding': \"Early folding (EFoldMine)\" \n",
        "}\n",
        "\n",
        "PREDICTION_POSITION = {\n",
        "    'backbone':     (0, 0),\n",
        "    'sidechain':    (0, 1),\n",
        "    'ppII':         (0, 2),\n",
        "    'coil':         (0, 3),\n",
        "    'sheet':        (1, 0),\n",
        "    'helix':        (1, 1),\n",
        "    'disoMine':     (1, 2),\n",
        "    'earlyFolding': (1, 3) \n",
        "}\n",
        "\n",
        "AXIS_TITLES = {\n",
        "    \"x\": \"Residue position in the MSA\",\n",
        "    \"y\": \"Prediction values\"\n",
        "}"
      ],
      "metadata": {
        "cellView": "form",
        "id": "G2sY5W-Op7ZB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 5.a Plot global results\n",
        "#@markdown Once this cell has been executed, a series of plots\n",
        "#@markdown with the distributions will be render.\n",
        "\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "def plot_msa_distrib_no_entropy(jsondata_list, sequences, mutation=False):\n",
        "    colors = ['blue', 'orange']\n",
        "    sequences_count = len(sequences)\n",
        "    residues_count = len(jsondata_list[0]['backbone']['median'])\n",
        "\n",
        "    #Plot representation\n",
        "    fig, axs = plt.subplots(2, 4)\n",
        "    fig.set_figwidth(30)\n",
        "    fig.set_figheight(10)\n",
        "    plt.subplots_adjust(hspace=0.4)\n",
        "\n",
        "    # These for loops got too complicated, I have to think\n",
        "    # something simpler to handle the None values in the data\n",
        "    predictions = jsondata_list[0].keys()\n",
        "    for prediction_index, biophys_data in enumerate(predictions):\n",
        "        if biophys_data == 'agmata':\n",
        "            continue\n",
        "        \n",
        "        subplot_index_row, subplot_index_col = PREDICTION_POSITION[biophys_data]\n",
        "\n",
        "        ax = axs[subplot_index_row, subplot_index_col]\n",
        "        for data, col in zip(jsondata_list, colors):\n",
        "            none_idx = []\n",
        "            \n",
        "            for n in range(residues_count):\n",
        "                if data[biophys_data]['median'][n] == None \\\n",
        "                        or data[biophys_data][\n",
        "                    'firstQuartile'][n] == None \\\n",
        "                        or data[biophys_data][\n",
        "                    'thirdQuartile'][n] == None:\n",
        "                    none_idx.append(n)\n",
        "\n",
        "            range_list = []\n",
        "            for n in range(len(none_idx)):\n",
        "                try:\n",
        "                    if none_idx[n] + 1 != none_idx[n + 1]:\n",
        "                        range_list.append(\n",
        "                            (none_idx[n] + 1, none_idx[n + 1]))\n",
        "                    else:\n",
        "                        continue\n",
        "                except:\n",
        "                    if len(none_idx) == 1:\n",
        "                        range_list.append((0, none_idx[0]))\n",
        "                        range_list.append((none_idx[0] + 1, len(\n",
        "                            data[biophys_data][\n",
        "                                'median'])))\n",
        "\n",
        "                    else:\n",
        "                        range_list.append((0, none_idx[0]))\n",
        "                        range_list.append((none_idx[-1] + 1, len(\n",
        "                            data[biophys_data][\n",
        "                                'median'])))\n",
        "\n",
        "            # When there are None values in the data\n",
        "            if range_list:\n",
        "                for tuple in range_list:\n",
        "                    x = np.arange(tuple[0], tuple[1], 1)\n",
        "                    firstq = \\\n",
        "                        data[biophys_data][\n",
        "                            'firstQuartile'][\n",
        "                        tuple[0]:tuple[1]]\n",
        "                    thirdq = \\\n",
        "                        data[biophys_data][\n",
        "                            'thirdQuartile'][\n",
        "                        tuple[0]:tuple[1]]\n",
        "                    bottom = \\\n",
        "                        data[biophys_data][\n",
        "                            'bottomOutlier'][\n",
        "                        tuple[0]:tuple[1]]\n",
        "                    top = \\\n",
        "                        data[biophys_data]['topOutlier'][\n",
        "                        tuple[0]:tuple[1]]\n",
        "                    ax.fill_between(\n",
        "                        x, firstq, thirdq, alpha=0.5, color=col, label=\"1st\")\n",
        "                    ax.fill_between(\n",
        "                        x, bottom, top, alpha=0.25, color=col, label=\"3rd\")\n",
        "\n",
        "            # When there aren't None values in the data\n",
        "            else:\n",
        "                x = np.arange(0, len(\n",
        "                    data[biophys_data]['median']), 1)\n",
        "                firstq = data[biophys_data][\n",
        "                    'firstQuartile']\n",
        "                thirdq = data[biophys_data][\n",
        "                    'thirdQuartile']\n",
        "                bottom = data[biophys_data][\n",
        "                    'bottomOutlier']\n",
        "                top = data[biophys_data]['topOutlier']\n",
        "                ax.fill_between(\n",
        "                    x, firstq, thirdq, alpha=0.5, color=col, label=\"1st\")\n",
        "                ax.fill_between(\n",
        "                    x, bottom, top, alpha=0.25, color=col, label=\"3rd\")\n",
        "\n",
        "            ax.plot(data[biophys_data]['median'], linewidth=1, color=col)\n",
        "\n",
        "            if mutation:\n",
        "                print(mutation)\n",
        "                ax.plot(mutation['results'][biophys_data],\n",
        "                         linewidth=0.5, color='red')\n",
        "\n",
        "        ax.set_title(PREDICTION_TITLES[biophys_data])\n",
        "        ax.axis([0, residues_count, -1.0, 1.5])\n",
        "        ax.set_ylabel(AXIS_TITLES['y'])\n",
        "        ax.set_xlabel(AXIS_TITLES['x'])\n",
        "    \n",
        "    return fig, axs\n",
        "  \n",
        "jsondata_list = [msaSuite.alignedPredictionDistribs]\n",
        "residues_count = len(jsondata_list[0]['backbone']['median'])\n",
        "sequences =  msaSuite.seqs\n",
        "fig, axs = plot_msa_distrib_no_entropy(jsondata_list, sequences, mutation=False)\n",
        "\n",
        "plt.suptitle(f\"MSA predicted biophysical properties: {residues_count} aligned residues from {len(sequences)} sequences\", fontsize=14)\n",
        "plt.legend(loc='center left', bbox_to_anchor=(1, 0.5), title = \"Quartiles:\")\n",
        "plt.tight_layout()\n",
        "plt.savefig('/content/results/msa.png' )\n",
        "plt.show()"
      ],
      "metadata": {
        "cellView": "form",
        "id": "EE_rrAuCpjIX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Define the ranges of residues to plot:\n",
        "\n",
        "all_regions = [\n",
        "    (\"First quarter\", (0, math.floor(residues_count * 0.25))),\n",
        "    (\"Second quarter\", (math.floor(residues_count * 0.25), math.floor(residues_count * 0.50))),\n",
        "    (\"Third quarter\", (math.floor(residues_count * 0.5), math.floor(residues_count * 0.75))),\n",
        "    (\"Fourth quarter\", (math.floor(residues_count * 0.75), residues_count)),\n",
        "    # (\"All\", (0, residues_count)),\n",
        "]"
      ],
      "metadata": {
        "id": "-UrXmWllrsWe"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 5.b. Plot the sequence conservation (Shannon's entropy) results\n",
        "#@markdown Once this cell has been executed, a plot\n",
        "#@markdown with the distributions will be render.\n",
        "\n",
        "import numpy as np\n",
        "import matplotlib as mpl\n",
        "import matplotlib.pyplot as plt\n",
        "from matplotlib.pyplot import figure\n",
        "\n",
        "def plot_msa_distrib_all(jsondata_list, sequences, all_regions, conservation_AA_list,mutation=False):\n",
        "    sequences_count = len(sequences)\n",
        "\n",
        "    #Color map for conservation sequence\n",
        "    cmap = mpl.cm.Blues\n",
        "    normalize = mpl.colors.Normalize(vmin=min(conservation_AA_list), vmax=max(conservation_AA_list))\n",
        "\n",
        "    colors = ['blue', 'orange']\n",
        "    for region in all_regions: \n",
        "\n",
        "        #To improve: use probabilistic description TM regions/ ICLs/ ECLs\n",
        "        name_region = region[0]\n",
        "        lower_lim = region[1][0]\n",
        "        upper_lim = region[1][1]\n",
        "\n",
        "        #Plot representation\n",
        "        fig = plt.figure()\n",
        "\n",
        "        ax1 = fig.add_subplot(241)\n",
        "        ax2 = fig.add_subplot(242)\n",
        "        ax3 = fig.add_subplot(243)\n",
        "        ax4 = fig.add_subplot(244)\n",
        "        ax5 = fig.add_subplot(245)\n",
        "        ax6 = fig.add_subplot(246)\n",
        "        ax7 = fig.add_subplot(247)\n",
        "        ax8 = fig.add_subplot(248)\n",
        "\n",
        "        ax_list = [ax1, ax2, ax3, ax4, ax5, ax6, ax7, ax8]\n",
        "\n",
        "        fig.set_figwidth(30)\n",
        "        fig.set_figheight(10)\n",
        "        fig.suptitle(f'MSA predicted biophysical properties: \"{name_region}\" (residue {lower_lim} to residue {upper_lim}) from {sequences_count} sequences', fontsize=14)\n",
        "        \n",
        "        plt.subplots_adjust(hspace=0.4)\n",
        "\n",
        "        #For every predicted biophysical property\n",
        "        for j, biophys_data in enumerate(jsondata_list[0].keys()):\n",
        "            if biophys_data == 'agmata':\n",
        "                continue\n",
        "            \n",
        "            for data in jsondata_list:\n",
        "                none_idx = []\n",
        "                for n in range(len(data[biophys_data]['median'])):\n",
        "                    if data[biophys_data]['median'][n] == None \\\n",
        "                            or data[biophys_data][\n",
        "                        'firstQuartile'][n] == None \\\n",
        "                            or data[biophys_data][\n",
        "                        'thirdQuartile'][n] == None:\n",
        "                        none_idx.append(n)\n",
        "\n",
        "                # # When there aren't None values in the data\n",
        "                # else:\n",
        "                x = np.arange(lower_lim,upper_lim, 1)\n",
        "                firstq = data[biophys_data][\n",
        "                    'firstQuartile'][lower_lim:upper_lim]\n",
        "                thirdq = data[biophys_data][\n",
        "                    'thirdQuartile'][lower_lim:upper_lim]\n",
        "                bottom = data[biophys_data][\n",
        "                    'bottomOutlier'][lower_lim:upper_lim]\n",
        "                top = data[biophys_data]['topOutlier'][lower_lim:upper_lim]\n",
        "                entropy_values = conservation_AA_list[lower_lim:upper_lim]\n",
        "                \n",
        "                for i in range(len(x)-1):\n",
        "                    ax_list[j].fill_between([x[i], x[i+1]], [firstq[i], firstq[i+1]], [thirdq[i], thirdq[i+1]], \\\n",
        "                        color=cmap(normalize(entropy_values[i])))                        \n",
        "                \n",
        "                ax_list[j].plot(x,firstq, linewidth=0.25, color=\"black\")\n",
        "                ax_list[j].plot(x,thirdq, linewidth=0.25, color=\"black\")\n",
        "                ax_list[j].plot(x,bottom, alpha=0.25, color=\"black\")\n",
        "                ax_list[j].plot(x,top, alpha=0.25, color=\"black\")\n",
        "\n",
        "                ax_list[j].plot(data[biophys_data]['median'],linewidth=1, color=\"black\")\n",
        "\n",
        "                if biophys_data == \"backbone\": \n",
        "                    ax1.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                    ax1.set_xlim([lower_lim, upper_lim-1])\n",
        "                    ax1.set_ylim([-0.2, 1.2])\n",
        "                    ax1.set_xlabel(AXIS_TITLES['x'])\n",
        "                    ax1.set_ylabel(AXIS_TITLES['y'])\n",
        "                    ax1.axhline(y=1.0, color='orange', linestyle='-', label=\"Membrane spaning\") #Membrane spaning\n",
        "                    ax1.axhline(y=0.8, color='blue', linestyle='-', label=\"Context dependent\") #context dependent (either rigide or flexible)\n",
        "                    ax1.axhline(y=0.69, color='red', linestyle='-', label=\"Flexible\") #flexible\n",
        "                    ax1.legend(title = \"Regions:\")\n",
        "\n",
        "                elif biophys_data == \"sidechain\":\n",
        "                    ax2.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                    ax2.set_xlim([lower_lim,upper_lim-1])\n",
        "                    ax2.set_ylim([-0.2, 1.2])\n",
        "                    ax2.set_xlabel(AXIS_TITLES['x'])\n",
        "                    ax2.set_ylabel(AXIS_TITLES['y'])\n",
        "\n",
        "                elif biophys_data == \"ppII\":\n",
        "                    ax3.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                    ax3.set_xlim([lower_lim,upper_lim-1])\n",
        "                    ax3.set_ylim([-0.2, 1.2])\n",
        "                    ax3.set_xlabel(AXIS_TITLES['x'])\n",
        "                    ax3.set_ylabel(AXIS_TITLES['y'])\n",
        "\n",
        "                elif biophys_data == \"coil\":\n",
        "                    ax4.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                    ax4.set_xlim([lower_lim,upper_lim-1])\n",
        "                    ax4.set_ylim([-0.2, 1.2])\n",
        "                    ax4.set_xlabel(AXIS_TITLES['x'])\n",
        "                    ax4.set_ylabel(AXIS_TITLES['y'])\n",
        "\n",
        "                elif biophys_data == \"sheet\":\n",
        "                    ax5.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                    ax5.set_xlim([lower_lim,upper_lim-1])\n",
        "                    ax5.set_ylim([-0.2, 1.2])\n",
        "                    ax5.set_xlabel(AXIS_TITLES['x'])\n",
        "                    ax5.set_ylabel(AXIS_TITLES['y'])\n",
        "\n",
        "                elif biophys_data ==\"helix\":\n",
        "                    ax6.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                    ax6.set_xlim([lower_lim,upper_lim-1])\n",
        "                    ax6.set_ylim([-0.2, 1.2])\n",
        "                    ax6.set_xlabel(AXIS_TITLES['x'])\n",
        "                    ax6.set_ylabel(AXIS_TITLES['y'])\n",
        "\n",
        "                elif biophys_data == \"earlyFolding\":\n",
        "                    ax7.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                    ax7.set_xlim([lower_lim,upper_lim-1])\n",
        "                    ax7.set_ylim([-0.2, 1.2])\n",
        "                    ax7.set_xlabel(AXIS_TITLES['x'])\n",
        "                    ax7.set_ylabel(AXIS_TITLES['y'])\n",
        "                    ax7.axhline(y=0.169, color='red', linestyle='-', label=\"above: Likely\") #above: likely start protein folding process\n",
        "                    ax7.legend(title = \"Regions:\")\n",
        "\n",
        "                elif biophys_data == \"disoMine\":\n",
        "                    ax8.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                    ax8.set_xlim([lower_lim,upper_lim-1])\n",
        "                    ax8.set_ylim([-0.2, 1.2])\n",
        "                    ax8.set_xlabel(AXIS_TITLES['x'])\n",
        "                    ax8.set_ylabel(AXIS_TITLES['y'])\n",
        "                    ax8.axhline(y=0.5, color='red', linestyle='-', label=\"above: Likely\") #above: likely disordered   \n",
        "                    ax8.legend(title = \"Regions:\")\n",
        "\n",
        "\n",
        "        #To add legend color map (where to put it???)\n",
        "        # scalarmappaple = mpl.cm.ScalarMappable(norm=normalize, cmap=cmap)\n",
        "        # scalarmappaple.set_array(entropy_values)\n",
        "        # plt.colorbar(scalarmappaple)\n",
        "\n",
        "        plt.savefig('/content/results/'+ name_region + '_entropy.png')\n",
        "        plt.tight_layout()\n",
        "        plt.show()\n",
        "  \n",
        "jsondata_list = [msaSuite.alignedPredictionDistribs]\n",
        "sequences = msaSuite.seqs\n",
        "residues_count = len(jsondata_list[0]['backbone']['median'])\n",
        "\n",
        "plot_msa_distrib_all(jsondata_list, sequences, all_regions, conservation_AA_list, mutation=False)"
      ],
      "metadata": {
        "cellView": "form",
        "id": "8YTLpmnqgTgZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 5.c. Plot the sequence conservation (Shannon's entropy) results (SPECIFIC REGION)\n",
        "#@markdown Once this cell has been executed, a plot\n",
        "#@markdown with the distributions will be render. \n",
        "\n",
        "#@markdown **Notes**: Use the blue bar on the bottom of the plots to adjust the region to render \n",
        "\n",
        "\n",
        "import numpy as np\n",
        "import matplotlib as mpl\n",
        "import matplotlib.pyplot as plt\n",
        "from matplotlib.widgets import Button, RangeSlider\n",
        "# from mpl_interactions.widgets import RangeSlider\n",
        "\n",
        "def plot_msa_distrib(jsondata_list, sequences, lower_lim, upper_lim, conservation_AA_list,mutation=False):\n",
        "    sequences_count = len(sequences)\n",
        "\n",
        "    #Color map for conservation sequence\n",
        "    cmap = mpl.cm.Blues\n",
        "    normalize = mpl.colors.Normalize(vmin=min(conservation_AA_list), vmax=max(conservation_AA_list))\n",
        "    colors = ['blue', 'orange']\n",
        "    \n",
        "    #To improve: use probabilistic description TM regions/ ICLs/ ECLs\n",
        "    name_region = 'section'\n",
        "\n",
        "    #Plot representation\n",
        "    fig = plt.figure()\n",
        "\n",
        "    ax1 = fig.add_subplot(241)\n",
        "    ax2 = fig.add_subplot(242)\n",
        "    ax3 = fig.add_subplot(243)\n",
        "    ax4 = fig.add_subplot(244)\n",
        "    ax5 = fig.add_subplot(245)\n",
        "    ax6 = fig.add_subplot(246)\n",
        "    ax7 = fig.add_subplot(247)\n",
        "    ax8 = fig.add_subplot(248)\n",
        "\n",
        "    ax_list = [ax1, ax2, ax3, ax4, ax5, ax6, ax7, ax8]\n",
        "\n",
        "    fig.set_figwidth(30)\n",
        "    fig.set_figheight(10)\n",
        "    plt.subplots_adjust(hspace=0.4)\n",
        "\n",
        "    fig.suptitle(f'MSA predicted biophysical properties: residue {int(lower_lim)} to residue {int(upper_lim)} from {sequences_count} sequences', fontsize=14)\n",
        "    \n",
        "    #For every predicted biophysical property\n",
        "    for j, biophys_data in enumerate(jsondata_list[0].keys()):\n",
        "        if biophys_data == 'agmata':\n",
        "            continue\n",
        "        \n",
        "        for data in jsondata_list:\n",
        "            none_idx = []\n",
        "            for n in range(len(data[biophys_data]['median'])):\n",
        "                if data[biophys_data]['median'][n] == None \\\n",
        "                        or data[biophys_data][\n",
        "                    'firstQuartile'][n] == None \\\n",
        "                        or data[biophys_data][\n",
        "                    'thirdQuartile'][n] == None:\n",
        "                    none_idx.append(n)\n",
        "\n",
        "            # # When there aren't None values in the data\n",
        "            # else:\n",
        "            x = np.arange(lower_lim,upper_lim, 1)\n",
        "            firstq = data[biophys_data][\n",
        "                'firstQuartile'][lower_lim:upper_lim]\n",
        "            thirdq = data[biophys_data][\n",
        "                'thirdQuartile'][lower_lim:upper_lim]\n",
        "            bottom = data[biophys_data][\n",
        "                'bottomOutlier'][lower_lim:upper_lim]\n",
        "            top = data[biophys_data]['topOutlier'][lower_lim:upper_lim]\n",
        "            entropy_values = conservation_AA_list[lower_lim:upper_lim]\n",
        "            \n",
        "            for i in range(len(x)-1):\n",
        "                ax_list[j].fill_between([x[i], x[i+1]], [firstq[i], firstq[i+1]], [thirdq[i], thirdq[i+1]], \\\n",
        "                    color=cmap(normalize(entropy_values[i])))                        \n",
        "            \n",
        "            ax_list[j].plot(x,firstq, linewidth=0.25, color=\"black\")\n",
        "            ax_list[j].plot(x,thirdq, linewidth=0.25, color=\"black\")\n",
        "            ax_list[j].plot(x,bottom, alpha=0.25, color=\"black\")\n",
        "            ax_list[j].plot(x,top, alpha=0.25, color=\"black\")\n",
        "\n",
        "            ax_list[j].plot(data[biophys_data]['median'],linewidth=1, color=\"black\")\n",
        "\n",
        "            if biophys_data == \"backbone\": \n",
        "                ax1.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                ax1.set_xlim([lower_lim, upper_lim-1])\n",
        "                ax1.set_ylim([-0.2, 1.2])\n",
        "                ax1.set_xlabel(AXIS_TITLES['x'])\n",
        "                ax1.set_ylabel(AXIS_TITLES['y'])\n",
        "                ax1.axhline(y=1.0, color='orange', linestyle='-', label=\"Membrane spaning\") #Membrane spaning\n",
        "                ax1.axhline(y=0.8, color='blue', linestyle='-', label=\"Context dependent\") #context dependent (either rigide or flexible)\n",
        "                ax1.axhline(y=0.69, color='red', linestyle='-', label=\"Flexible\") #flexible\n",
        "                ax1.legend(title = \"Regions:\")\n",
        "\n",
        "            elif biophys_data == \"sidechain\":\n",
        "                ax2.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                ax2.set_xlim([lower_lim,upper_lim-1])\n",
        "                ax2.set_ylim([-0.2, 1.2])\n",
        "                ax2.set_xlabel(AXIS_TITLES['x'])\n",
        "                ax2.set_ylabel(AXIS_TITLES['y'])\n",
        "\n",
        "            elif biophys_data == \"ppII\":\n",
        "                ax3.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                ax3.set_xlim([lower_lim,upper_lim-1])\n",
        "                ax3.set_ylim([-0.2, 1.2])\n",
        "                ax3.set_xlabel(AXIS_TITLES['x'])\n",
        "                ax3.set_ylabel(AXIS_TITLES['y'])\n",
        "\n",
        "            elif biophys_data == \"coil\":\n",
        "                ax4.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                ax4.set_xlim([lower_lim,upper_lim-1])\n",
        "                ax4.set_ylim([-0.2, 1.2])\n",
        "                ax4.set_xlabel(AXIS_TITLES['x'])\n",
        "                ax4.set_ylabel(AXIS_TITLES['y'])\n",
        "\n",
        "            elif biophys_data == \"sheet\":\n",
        "                ax5.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                ax5.set_xlim([lower_lim,upper_lim-1])\n",
        "                ax5.set_ylim([-0.2, 1.2])\n",
        "                ax5.set_xlabel(AXIS_TITLES['x'])\n",
        "                ax5.set_ylabel(AXIS_TITLES['y'])\n",
        "\n",
        "            elif biophys_data ==\"helix\":\n",
        "                ax6.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                ax6.set_xlim([lower_lim,upper_lim-1])\n",
        "                ax6.set_ylim([-0.2, 1.2])\n",
        "                ax6.set_xlabel(AXIS_TITLES['x'])\n",
        "                ax6.set_ylabel(AXIS_TITLES['y'])\n",
        "\n",
        "            elif biophys_data == \"earlyFolding\":\n",
        "                ax7.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                ax7.set_xlim([lower_lim,upper_lim-1])\n",
        "                ax7.set_ylim([-0.2, 1.2])\n",
        "                ax7.set_xlabel(AXIS_TITLES['x'])\n",
        "                ax7.set_ylabel(AXIS_TITLES['y'])\n",
        "                ax7.axhline(y=0.169, color='red', linestyle='-', label=\"above: Likely\") #above: likely start protein folding process\n",
        "                ax7.legend(title = \"Regions:\")\n",
        "\n",
        "            elif biophys_data == \"disoMine\":\n",
        "                ax8.set_title(PREDICTION_TITLES[biophys_data])\n",
        "                ax8.set_xlim([lower_lim,upper_lim-1])\n",
        "                ax8.set_ylim([-0.2, 1.2])\n",
        "                ax8.set_xlabel(AXIS_TITLES['x'])\n",
        "                ax8.set_ylabel(AXIS_TITLES['y'])\n",
        "                ax8.axhline(y=0.5, color='red', linestyle='-', label=\"above: Likely\") #above: likely disordered   \n",
        "                ax8.legend(title = \"Regions:\")\n",
        "\n",
        "    return fig, ax_list\n",
        "\n",
        "\n",
        "jsondata_list = [msaSuite.alignedPredictionDistribs]\n",
        "sequences = msaSuite.seqs\n",
        "residues_count = len(jsondata_list[0]['backbone']['median'])\n",
        "\n",
        "# Create the figure and the line that we will manipulate\n",
        "fig, ax_list = plot_msa_distrib(jsondata_list, sequences, 0, residues_count, conservation_AA_list)\n",
        "\n",
        "# Make a horizontal slider to control the frequency.\n",
        "ax_residues_slider = plt.axes([0.25, 0.1, 0.65, 0.03])\n",
        "init = (0, residues_count)\n",
        "residues_slider = RangeSlider(ax_residues_slider, 'Residues', 0, residues_count, valstep=1)\n",
        "\n",
        "# The function to be called anytime a slider's value changes\n",
        "def update(val):\n",
        "    new_lower_lim, new_upper_lim = residues_slider.val\n",
        "    for ax in ax_list:\n",
        "        # line.set_xdata(np.linspace(residues_slider.val[0], residues_slider.val[1], 100))\n",
        "        ax.set_xlim(new_lower_lim, new_upper_lim)\n",
        "        # line.set_ydata(reliability_graph(t, residues_slider.val[0], residues_slider.val[1]))\n",
        "    \n",
        "    fig.canvas.draw_idle()\n",
        "    fig.suptitle(f'MSA predicted biophysical properties: residue {new_lower_lim} to residue {new_upper_lim} from {len(sequences)} sequences')\n",
        "\n",
        "# register the update function with each slider\n",
        "residues_slider.on_changed(update)\n",
        "\n",
        "# adjust the main plot to make room for the sliders\n",
        "plt.tight_layout()\n",
        "plt.subplots_adjust(left=0.05, bottom=0.25)\n",
        "plt.show()"
      ],
      "metadata": {
        "cellView": "form",
        "id": "63B3qptOpeeZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 6. Save the predictions to json files\n",
        "\n",
        "#@markdown Once this cell has been executed, a json file with\n",
        "#@markdown the obtained predictions will be automatically created for each\n",
        "#@markdown sequence of the alignment inside the /content/results folder\n",
        "%%capture\n",
        "\n",
        "import os\n",
        "from os import path\n",
        "\n",
        "if not path.exists(\"/content/results\"):\n",
        "    os.mkdir(\"/content/results\", )\n",
        "\n",
        "for id, seq in msaSuite.seqs:\n",
        "    print(\"PROCESSING SEQUENCE: id={id}\".format(id=id))\n",
        "\n",
        "    json_results = msaSuite.getAllPredictionsJson(identifier=id)\n",
        "    # print(json_results)\n",
        "    with open(\"/content/results/{msa}_{id}.json\".format(msa=msa_filename.replace(\"/\", \"_\"), id=id.replace(\"/\", \"_\")), \"w\") as file_output:\n",
        "        file_output.write(json_results)\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "5zgf8PmCb717"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 7. Download the predictions\n",
        "\n",
        "#@markdown Once this cell has been executed, a zip-archive with\n",
        "#@markdown the obtained predictions will be automatically downloaded\n",
        "#@markdown to your computer.\n",
        "\n",
        "!zip --quiet -r /content/b2b-tools-msa.zip /content/results\n",
        "\n",
        "files.download(f\"/content/b2b-tools-msa.zip\")"
      ],
      "metadata": {
        "cellView": "form",
        "id": "Nckxe9SPemTr"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}